its-a-smol-world/src/smol_mind.py

```diff
@@ -0,0 +1,167 @@
+    You are an expert designed to call the correct function to solve a problem based on the user's request.
+    The functions available (with required parameters) to you are:
+    {functions}
+    
+    You will be given a user prompt and you need to decide which function to call.
+    You will then need to format the function call correctly and return it in the correct format.
+    The format for the function call is:
+    [func1(params_name=params_value]
+    NO other text MUST be included.
+                                      
+    For example:
+    Request: I want to order a cheese pizza from Pizza Hut.
+    Response: [order_food(restaurant="Pizza Hut", item="cheese pizza", quantity=1)]
+                                      
+    Request: I want to know the weather in Tokyo.
+    Response: [get_weather(city="Tokyo")]
+
+    Request: I need a ride to SFO.
+    Response: [order_ride(destination="SFO")]
+""")
+
+
+ASSISTANT_PROMPT_FOR_CHAT_MODEL = dedent("""
+    I understand and will only return the function call in the correct format.
+    """
+)
+USER_PROMPT_FOR_CHAT_MODEL = dedent("""
+    Request: {user_prompt}. 
+""")
+
+
+def instruct_prompt(question, functions, tokenizer):
+    messages = [
+        {"role": "user", "content": SYSTEM_PROMPT_FOR_CHAT_MODEL.format(functions=format_functions(functions))},
+        {"role": "assistant", "content": ASSISTANT_PROMPT_FOR_CHAT_MODEL },
+        {"role": "user", "content": USER_PROMPT_FOR_CHAT_MODEL.format(user_prompt=question)},
+    ]
+    fc_prompt = tokenizer.apply_chat_template(messages, tokenize=False)
+    return fc_prompt
+
+INTEGER = r"(-)?(0|[1-9][0-9]*)"
+STRING_INNER = r'([^"\\\x00-\x1F\x7F-\x9F]|\\["\\])'
+# We'll limit this to just a max of 42 characters
+STRING = f'"{STRING_INNER}{{1,42}}"'
+# i.e. 1 is a not a float but 1.0 is.
+FLOAT = rf"({INTEGER})(\.[0-9]+)([eE][+-][0-9]+)?"
+BOOLEAN = r"(true|false)"
+NULL = r"null"
+
+simple_type_map = {
+    "string": STRING,
+    "any": STRING,
+    "integer": INTEGER,
+    "number": FLOAT,
+    "float": FLOAT,
+    "boolean": BOOLEAN,
+    "null": NULL,
+}
+
+def build_dict_regex(props):
+    out_re = r"\{"
+    args_part = ", ".join(
+        [f'"{prop}": ' + type_to_regex(props[prop]) for prop in props]
+    )
+    return out_re + args_part + r"\}"
+
+def type_to_regex(arg_meta):
+    arg_type = arg_meta["type"]
+    if arg_type == "object":
+        arg_type = "dict"
+    if arg_type == "dict":
+        try:
+            result = build_dict_regex(arg_meta["properties"])
+        except KeyError:
+            return "Definition does not contain 'properties' value."
+    elif arg_type in ["array","tuple"]:
+        pattern = type_to_regex(arg_meta["items"])
+        result = r"\[(" + pattern + ", ){0,8}" + pattern + r"\]"
+    else:
+        result = simple_type_map[arg_type]
+    return result
+
+type_to_regex({
+    "type": "array",
+    "items": {"type": "float"}
+})
+
+def build_standard_fc_regex(function_data):
```

Comment:
Would you mind documenting these functions? I opened [an issue](https://github.com/dottxt-ai/outlines/issues/1217) on Outlines to provide an interface for multi-functions function calling and we will probably use this code as a first draft.

Reply:
Absolutely!